tool/tsh/kube.go

```diff
@@ -270,6 +270,90 @@ func fetchKubeClusters(ctx context.Context, tc *client.TeleportClient) (teleport
 	return teleportCluster, kubeClusters, nil
 }
 
+// kubernetesStatus holds teleport client information necessary to populate the user's kubeconfig.
+type kubernetesStatus struct {
+	clusterAddr         string
+	teleportClusterName string
+	kubeClusters        []string
+	credentials         *client.Key
+}
+
+// fetchKubeStatus returns a kubernetesStatus populated from the given TeleportClient.
+func fetchKubeStatus(ctx context.Context, tc *client.TeleportClient) (*kubernetesStatus, error) {
+	var err error
+	kubeStatus := &kubernetesStatus{
+		clusterAddr: tc.KubeClusterAddr(),
+	}
+	kubeStatus.credentials, err = tc.LocalAgent().GetCoreKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	kubeStatus.teleportClusterName, kubeStatus.kubeClusters, err = fetchKubeClusters(ctx, tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return kubeStatus, nil
+}
+
+// buildKubeConfigUpdate returns a kubeconfig.Values suitable for updating the user's kubeconfig
+// based on the CLI parameters and the given kubernetesStatus.
+func buildKubeConfigUpdate(cf *CLIConf, kubeStatus *kubernetesStatus) (*kubeconfig.Values, error) {
+	v := &kubeconfig.Values{
+		ClusterAddr:         kubeStatus.clusterAddr,
+		TeleportClusterName: kubeStatus.teleportClusterName,
+		Credentials:         kubeStatus.credentials,
+	}
+
+	if cf.executablePath != "" {
+		v.Exec = &kubeconfig.ExecValues{
+			TshBinaryPath:     cf.executablePath,
+			TshBinaryInsecure: cf.InsecureSkipVerify,
+			KubeClusters:      kubeStatus.kubeClusters,
+		}
+
+		// Only switch the current context if kube-cluster is explicitly set on the command line.
+		if cf.KubernetesCluster != "" {
+			if !utils.SliceContainsStr(kubeStatus.kubeClusters, cf.KubernetesCluster) {
+				return nil, trace.BadParameter("kubernetes cluster %q is not registered in this teleport cluster; you can list registered kubernetes clusters using 'tsh kube ls'", cf.KubernetesCluster)
```

Comment:
Nit: Let's capitalize Kubernetes and Teleport consistently (they're capitalized in message below).